File: LeetCode/src/0284 - Peeking Iterator.java
// Java Iterator interface reference:
// https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html

class PeekingIterator implements Iterator<Integer> {
    private Iterator<Integer> _iterator;
    private Integer _nextValue;

    public PeekingIterator(Iterator<Integer> iterator) {
        _iterator = iterator;
        _nextValue = iterator.next();
    }

    public Integer peek() {
        return _nextValue;
    }

    @Override
    public Integer next() {
        Integer res = _nextValue;
        _nextValue = _iterator.hasNext() ? _iterator.next() : null;
        return res;
    }

    @Override
    public boolean hasNext() {
        return _nextValue != null;
    }
}

File: LeetCode/src/0284 - Peeking Iterator.cs
// C# IEnumerator interface reference:
// https://docs.microsoft.com/en-us/dotnet/api/system.collections.ienumerator?view=netframework-4.8

class PeekingIterator
{
    private IEnumerator<int> _iter;
    private bool _hasNext;

    public PeekingIterator(IEnumerator<int> iterator)
    {
        _iter = iterator;
        _hasNext = iterator.MoveNext();
    }

    public int Peek()
    {
        return _iter.Current;
    }

    public int Next()
    {
        int res = _iter.Current;
        _hasNext = _iter.MoveNext();
        return res;
    }

    public bool HasNext()
    {
        return _hasNext;
    }
}